File: 2020/5/5.pl
#!/usr/bin/env swipl
% usage: swipl 5.pl FILENAME
% https://adventofcode.com/2020/day/5

:- module(aoc2020day5, []).
:- module(aoc2020day5).
:- initialization(main, main).
:- use_module(library(dcg/basics)).
:- use_module(library(dcg/high_order)).

load(Filename, IDs) :- once(phrase_from_file(load(IDs), Filename)).
load(IDs) --> sequence(seating_plan(0), IDs0), { sort(0, @>, IDs0, IDs) }.

seating_plan(P0, P) --> ( "F" | "L" ), { P1 is 2 * P0 },     seating_plan(P1, P).
seating_plan(P0, P) --> ( "B" | "R" ), { P1 is 2 * P0 + 1 }, seating_plan(P1, P).
seating_plan(P,  P) --> "\n".

part_one([Result | _], Result).

part_two([S1, S0 | Ss], Seat) :-
    ( succ(S0, S1) -> part_two([S0 | Ss], Seat) ; succ(Seat, S1) ).

main([File]) :-
    load(File, IDs),
    part_one(IDs, ResultOne),
    writeln(ResultOne),
    part_two(IDs, ResultTwo),
    writeln(ResultTwo).

File: 2025/5/5.jl
#!/usr/bin/env julia
# usage: julia 5.jl [FILENAME]
# https://adventofcode.com/2025/day/5

module AOC2025Day5

function load(file::AbstractString)
    lines = eachline(file)
    ranges = [range(parse.(Int, split(line, '-'))...) for line in Iterators.takewhile(!isempty, lines)]
    ranges, parse.(Int, lines)
end

function partone((ranges, ids))
    sort!(ranges, by=last)
    count(ids) do id
        index = searchsortedfirst(ranges, id, by=last)
        while index in keys(ranges) && id <= last(ranges[index])
            if id in ranges[index]
                return true
            end
            index += 1
        end
        false
    end
end

function combineranges(ranges)
    sort!(ranges, by=first)
    result = []
    combined = popfirst!(ranges)
    for range in ranges
        if first(range) <= last(combined)
            combined = first(combined):max(last(combined), last(range))
        else
            push!(result, combined)
            combined = range
        end
    end
    push!(result, combined)
    result
end

parttwo = Base.Fix1(sum, length) ∘ combineranges ∘ first

if abspath(PROGRAM_FILE) == @__FILE__
    input = load(get(ARGS, 1, joinpath(@__DIR__, "input.txt")))
    println(partone(input))
    println(parttwo(input))
elseif isinteractive()
    using REPL
    REPL.activate(AOC2025Day5)
end

end # module
